File: Leetcode/Dynamic Programming/576. Out of Boundary Paths.java
class Solution {
    private int allPaths(int x, int y, int X, int Y, int moves, int[][][] paths, int[] ds) {
        if (x < 0 || y < 0 || x >= X || y >= Y) return 1;
        if (moves <= 0) return 0;
        if (paths[x][y][moves] == -1) {
            long ways = 0;
            for (int i = 0; i < 4; i++) {
                ways += allPaths(x + ds[i], y + ds[3 - i], X, Y, moves - 1, paths, ds);
                ways %= 1000000007;
            }
            paths[x][y][moves] = (int)ways;
        }
        return paths[x][y][moves];
    }
    public int findPaths(int m, int n, int maxMove, int startRow, int startColumn) {
        int[][][] paths = new int[m][n][maxMove + 1];
        int[] ds = {-1, 1, 0, 0};
        for (int[][] matrix: paths) {
            for (int[] row: matrix) {
                Arrays.fill(row, -1);
            }
        }
        return allPaths(startRow, startColumn, m, n, maxMove, paths, ds) % 1000000007;
    }
}

File: Leetcode/Dynamic Programming/576. Out of Boundary Paths.py
class Solution:
    def findPathsH(self, m, n, maxMove, i, j):
        if i >= m or i < 0 or j >= n or j < 0:
            return 1
        if maxMove <= 0:
            return 0

        if self.paths[i][j][maxMove] == -1:
            res = (
                self.findPathsH(m, n, maxMove - 1, i + 1, j)
                + self.findPathsH(m, n, maxMove - 1, i - 1, j)
                + self.findPathsH(m, n, maxMove - 1, i, j + 1)
                + self.findPathsH(m, n, maxMove - 1, i, j - 1)
            )
            self.paths[i][j][maxMove] = res % self.mod
            
        return self.paths[i][j][maxMove]

    def findPaths(
        self, m: int, n: int, maxMove: int, startRow: int, startColumn: int
    ) -> int:
        self.mod = 1000000007
        self.paths = [
            [[-1 for _ in range(maxMove + 1)] for _ in range(n)] for _ in range(m)
        ]
        return self.findPathsH(m, n, maxMove, startRow, startColumn) % self.mod
